index.js

```diff
@@ -7,555 +7,103 @@
- * that contain async ids. This function should be used
- * with [co][].
- *
- * This is used inside `resolveArgs`:
- *
- * ```js
- * var args = yield co(asyncHelpers.resolveObject(options.hash));
- * ```
- * @param {Object} `obj` object with with values that may be async ids.
- * @returns {Object} Object with resolved values.
- */
-
-AsyncHelpers.prototype.resolveObject = function * (obj) {
-  var keys = Object.keys(obj);
-  var self = this;
 
-  return yield keys.reduce(function(acc, key) {
-    return co(function * () {
-      var val = acc[key];
-      if (typeof val === 'string' && self.hasAsyncId(val)) {
-        acc[key] = yield self.resolveId(val);
-      }
-      return acc;
-    });
-  }, obj);
-};
-
-/**
- * After rendering a string using wrapped async helpers,
- * use `resolveIds` to invoke the original async helpers and replace
- * the async ids with results from the async helpers.
- *
- * ```js
- * asyncHelpers.resolveIds(renderedString, function(err, content) {
- *   if (err) return console.error(err);
- *   console.log(content);
- * });
- * ```
- * @param  {String} `str` String containing async ids
- * @param  {Function} `cb` Callback function accepting an `err` and `content` parameters.
- * @api public
- */
-
-AsyncHelpers.prototype.resolveIds = function(str, cb) {
-  if (typeof cb !== 'function') {
-    throw new TypeError('AsyncHelpers#resolveIds() expects a callback function.');
-  }
-  if (typeof str !== 'string') {
-    return cb(new TypeError('AsyncHelpers#resolveIds() expects a string.'));
+    return this;
   }
 
-  var matches = this.matches(str);
-  var self = this;
-
-  co(function * () {
-    if (!matches) {
-      return str;
-    };
-
-    for (var i = 0; i < matches.length; i++) {
-      var key = matches[i];
-      var val = yield self.resolveId(key);
-      str = str.split(key).join(val);
-    }
-    return str;
-  })
-  .then((res) => cb(null, res))
-  .catch(cb);
-};
-
-/**
- * Format an error message to provide better information about the
- * helper and the arguments passed to the helper when the error occurred.
- *
- * @param  {Object} `err` Error object
- * @param  {Object} `helper` helper object to provide more information
- * @param  {Array} `args` Array of arguments passed to the helper.
- * @return {Object} Formatted Error object
- */
-
-function formatError(err, helper, args) {
-  err.helper = helper;
-  define(err, 'args', args);
-  return err;
-}
-
-/**
- * Create a prefix to use when generating an async id.
- *
- * @param  {String} `prefix` prefix string to start with.
- * @param  {String} `counter` string to append.
- * @return {String} new prefix
- */
+  async resolve(str) {
+    const self = this;
+    const promises = Object.keys(this.ids)
+      .map((id) => this.ids[id])
+      .map(async({ id, args, fn, context }) => {
 
-function appendPrefix(prefix, counter) {
-  return prefix + counter + '$';
-}
+        const argz = args.map(function func(arg) {
```

Comment:
can we use fat arrow here and `this`?

Reply:
I'm using a regular one because later we `map` over the `item.args`, so to be recursive. And the context should be `item.context`.